src/server.rs

```diff
@@ -51,35 +71,40 @@ impl DataResponse {
     }
 }
 
-pub async fn metadata_handler(payload: actix_web::web::Bytes) -> actix_web::HttpResponse {
-    let payload_to_str = std::str::from_utf8(&payload);
-    let parse_result = match payload_to_str {
-        Ok(r) => json::parse(r),
-        Err(_err) => Err(json::Error::FailedUtf8Parsing),
-    };
-
-    let body = parse_result.unwrap_or_else(|e| json::object! { "error" => e.to_string() });
+pub async fn metadata_handler(
+    app_state: actix_web::web::Data<context::AppState>,
+    payload: actix_web::web::Json<MetadataMessage>,
+) -> actix_web::HttpResponse {
+    let mut server_ctx = app_state.0.lock().unwrap();
 
-    match serde_json::from_str::<'_, MetadataMessage>(&body.dump()) {
-        Ok(b) => actix_web::HttpResponse::Ok()
-            .content_type("application/json")
-            .body(
-                serde_json::to_string(&b)
-                    .unwrap_or_else(|_| "Failed to convert body to a valid string".to_string()),
-            ),
-        Err(e) => actix_web::HttpResponse::build(actix_web::http::StatusCode::BAD_REQUEST).json(
-            ErrorResponse {
-                error: e.to_string(),
-            },
-        ),
+    match payload.into_inner() {
+        MetadataMessage::TrackTable(table) => {
+            match (*server_ctx).metadata_track_table(table.clone()) {
+                Ok(_) => make_metadata_success_response(
+                    format!("{} is now being tracked!", table.to_string()).as_str(),
+                ),
+                Err(err) => bad_request_response(format!("{}", err).as_str()),
+            }
+        }
+        MetadataMessage::UntrackTable(table) => {
+            match (*server_ctx).metadata_untrack_table(table.clone()) {
+                Ok(_) => make_metadata_success_response(
+                    format!("{} has now been un-tracked!", table.to_string()).as_str(),
+                ),
+                Err(err) => bad_request_response(format!("{}", err).as_str()),
+            }
+        }
+        MetadataMessage::ExportMetadata => {
+            actix_web::web::HttpResponse::Ok().json((*server_ctx).get_metadata())
```

Comment:
Shamelessly adding clone here 😅
```suggestion
          MetadataResponse::Data((*server_ctx).get_metadata().clone())
```

Reply:
I don't think clone is ever a bad idea. We can always ask for help from the `--release` flag! :laughing: